File: day20/python/day20.py
def merge(ints):
    ints.sort()
    stack = [ints[0]]
    for start, end in ints[1:]:
        ts, te = stack[-1]
        if te <= start:
            stack.append((start, end))
        elif te < end:
            stack.pop()
            stack.append((ts, end))
    return stack

def day20(input):
    intervals = merge([tuple(map(int, l.split("-"))) for l in input])
    part1 = intervals[0][1] + 1
    part2 = sum(c - b - 1 for (a, b), (c, d) in zip(intervals, intervals[1:]))
    return part1, part2

input = open("../input.txt").read()
input = [x.strip() for x in input.split("\n")]
print(day20(input))

File: day20/haskell/day20.hs
import Control.Arrow ((***), (&&&))
import Data.List (sort, break)
import Data.Tuple (swap)

intervals :: [String] -> [(Int, Int)]
intervals = sort . map parse
    where parse = (read *** read . tail) . break (=='-')

merge :: [(Int, Int)] -> [(Int, Int)] -> [(Int, Int)]
merge xs []     = reverse xs
merge [] (y:ys) = merge [y] ys
merge (x@(xl, xh):xs) (y@(yl, yh):ys)
    | xh <= yl  = merge (y:x:xs) ys
    | xh <  yh  = merge ((xl, yh):xs) ys
    | otherwise = merge (x:xs) ys

solve :: [String] -> (Int, Int)
solve = (lowest &&& total) . merge [] . intervals
    where lowest = (+1) . snd . head
          total = sum . map f . (zip <*> tail)
          f = subtract 1 . uncurry (-) . swap . (snd *** fst)

main :: IO ()
main = do
    input <- lines <$> readFile "../input.txt"
    print . solve $ input